draft-gpew-priv-ppm.md

```diff
@@ -672,242 +679,366 @@ independently.
+~~~
+enum {
+  continued(0),
+  finished(1)
+  failed(2),
+} ProcessType;
 
-[OPEN ISSUE: consider updating the `checksum` algorithm to not permit
-collisions]
+struct {
+  Nonce nonce;
+  ProcessType process_type;
+  select (ProcessShare.process_type) {
+    case continued: opaque agg_msg<0..2^16-1>; // VDAF output message
+    case finished:  // empty
+    case failed:    ReportShareError;
+  }
+} ProcessShare;
 
-First, the leader computes a checksum over the set of output shares included in
-the batch window. The checksum is computed by taking the SHA256 hash of each
-nonce from the client reports included in the aggregation, then combining the
-hash values with a bitwise-XOR operation.
+struct {
+  opaque helper_state<0..2^16>;
+  ProcessShare seq<1..2^16-1>;
+} AggregateInitResp;
+~~~
 
-Then, for each aggregator endpoint `[aggregator]` in the parameters associated
-with `CollectReq.task_id` (see {{pa-collect}}) except its own, the leader sends
-a POST request to `[aggregator]/aggregate_share` with the following message:
+The `helper_state` parameter contains the helper's initial state. This is an
+optional parameter of an aggregate request that the helper can use to carry
+state across requests. At least part of the state will usually need to be
+encrypted in order to protect user privacy. However, the details of precisely
+how the state is encrypted and the information that it carries is up to the
+helper implementation.
+
+[[OPEN ISSUE: we may end up removing helper_state. See #185]]
+
+The rest of the message is a sequence of ProcessShare values, the order of which
+matches that of the InputReportShare values in `AggregateInitReq.seq`. Each report
+that was marked as invalid is assigned the ProcessType `failed`. Otherwise, the
+ProcessShare is either marked as continued with the output `agg_msg`, or is marked
+as finished if the VDAF computation is finished.
+
+The helper's response to the leader is an HTTP 200 OK whose body is the
+AggregateInitResp and media type is "message/ppm-aggregate-init-resp".
+
+Upon receipt of a helper's AggregateInitResp message, the leader checks that the
+sequence of ProcessShare messages corresponds to the ReportShare sequence of the
+AggregateInitReq. If any message appears out of order, is missing, has an
+unrecognized nonce, or if two messages have the same nonce, then the leader MUST
+abort with error "unrecognizedMessage".
+
+### Aggregate Continuation {#agg-continue-flow}
+
+In the continuation phase, the leader drives aggregation of each share in the candidate
+report set until the underlying VDAF moves into a terminal state, which happens either
+when aggregation completes or an error occurs. The leader does this by repeating the
+following procedure.
+
+* If the helper failed, then mark the report as failed, remove it from the candidate
+  report set, and do not process it further.
+* If the leader continued but the helper finished, or if the leader finished but
```

Comment:
Can we get a forward ref to how things ifnish.

Reply:
Yep, done.